File: c++/92_Reverse_Linked_List_II.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
    	if (m == n)
    		return head;
    	ListNode *p = new ListNode(0);
    	p->next = head;
    	head = p;
    	ListNode *left = p, *right = p;
    	for (int i = 0; i < m; i++) {
    		left = left->next;
    		if (i != m - 1)
    			p = p->next;
    	}
    	for (int i = 0; i < n; i++)
    		right = right->next;
    	p->next = right;
    	right = right->next;

    	for (int i = n - m; i > 0; i--) {
    		ListNode *tmp = left;
    		for (int j = i - 1; j > 0; j--)
    			tmp = tmp->next;
    		tmp->next->next = tmp;
    	}
    	left->next = right;
    	return head->next;   
    }
};


int main() {
	Solution s;
}

File: python/92_Reverse_Linked_List_II.py
#!/usr/bin/python
import unittest


class Solution(object):
    def reverseBetween(self, head, m, n):
    	if m == n:
    		return head
    	p = ListNode(0)
    	p.next = head
    	head = p
    	leftN, rightN = p, p
    	for i in range(m):
    		leftN = leftN.next
    		if i != m - 1:
    			p = p.next
    	for j in range(n):
    		rightN = rightN.next
    	p.next = rightN  # 1->4
    	rightN = rightN.next
    	for i in range(n - m, 0, -1):
    		tmp = leftN
    		for j in range(i - 1, 0, -1):
    			tmp = tmp.next
    		tmp.next.next = tmp
    	leftN.next = rightN # 2->5
    	return head.next
    	



class TestSolution(unittest.TestCase):
    def setUp(self):
        self.s = Solution()

    def test_method(self):
        """Such as self.assertEqual, self.assertTrue"""
        pass


if __name__ == "__main__":
    unittest.main()